File: Python/GSocket.py
import socket
import json
import random

Packets = json.loads(json.dumps({}))

global IP
global Port
IP = "localhost"
Port = 5000
def Socket():
	return socket.socket(socket.AF_INET, socket.SOCK_DGRAM)

def Listen():
	s = Socket()
	s.bind((IP, Port))
	while True:
		data, address = s.recvfrom(4096)
		try:
			data = json.loads(data.decode('utf-8'))
			Header = json.loads(data["Header"])
			if not data["Body"] or not Header:
				continue

			if len(data["Body"]) != Header["Length"]:
				print("Packet data has been lost!")

			Body = json.loads(data["Body"])

			if Packets[Header["Type"]] != None:
				Packets[Header["Type"]](Body)
		except ValueError as e:
			print("Invalid packet has beed received!")

	
def Send(Type, JSON):
	try:
		Body = str(JSON)

		Header = {
			"Type" : Type,
			"Length" : len(Body)
		}

		packet = json.dumps({
			"Header" : Header,
			"Body" : Body
		})

		S = Socket()
		S.bind((IP, Port + random.randint(10, 1000)))
		S.sendto(packet.encode('utf-8'), (IP, Port))
		return S
	except ValueError as e:
		print("Invalid JSON String!")

def Receive(Socket):
	Empty = True
	while Empty:
		data, address = Socket.recvfrom(4096)
		try:
			data = json.loads(data.decode('utf-8'))
			Header = json.loads(data["Header"])
			if len(data["Body"]) != Header["Length"]:
				print("Packet data has been lost!")

			Body = json.loads(data["Body"])
			if Packets[Header["Type"]] != None:
				Packets[Header["Type"]](Body)
				Empty = False
		except ValueError as e:
			print("Invalid Packet has beed received!")


def AddPacket(Name, func):
	Packets[Name] = func

File: NodeJS/GSocket.js
var udp = require("dgram")

class GSocket {
	constructor() {

	}

	isJsonString(str) {
	    try {
	    	console.log(str);
	        JSON.parse(str);
	    } catch (e) {
	        return false;
	    }
	    return true;
	}

	Connect(IP, Port) {
		this.Client = udp.createSocket("udp4");
		this.IP = IP || "localhost";
		this.Port = Port || 5000;
		this.Packets = {};
	}

	Send(Name, str) {
		console.log(typeof str);
		if(!(typeof str === 'object')) {
			if(this.isJsonString(str)) {
				str = JSON.parse(str);
			} else {
				throw new Error('Type is not a String!');
			}
		} else {
			str = JSON.stringify(str);
		}

		var packet = '{"Header": {"Type": "' + Name + '", "Length": ' + str.length + '}, "Body": ' + str + '}';
		this.Client.send(Buffer.from(packet), this.Port, this.IP, function(error) {
			if(error) {
				this.Client.close();
			}
		});
	}

	Listen() {
		this.Client.on('message', function(msg, info) {
			msg = msg.tostring()
			if(this.isJsonString(msg)) {
				if(msg["Header"] != null || msg["Header"]["Length"] != null || msg["Header"]["Type"] != null || msg["Body"] != null) {
					if(this.Packets[msg["Header"]["Type"]] != null) {
						this.Packets[msg["Header"]["Type"]](msg);
					}
				}
			}
		});

		this.Client.on('listening', function() {
			var address = this.Client.address();
			var port = address.port;
			var family = address.family;
			var ipaddr = address.address;
			console.log('Server is listening at port' + port);
			console.log('Server ip :' + ipaddr);
			console.log('Server is IP4/IP6 : ' + family);
		});

		this.Client.on('close', function() {
			console.Log("Server closed!");
		});
		this.Client.bind(this.Port);
	}

	Receive() {
		this.Client.on('message', function(msg, info){
			msg = msg.tostring()
			if(this.isJsonString(msg)) {
				if(msg["Header"] != null || msg["Header"]["Length"] != null || msg["Header"]["Type"] != null || msg["Body"] != null) {
					if(this.Packets[msg["Header"]["Type"]] != null) {
						this.Packets[msg["Header"]["Type"]](msg);
					}
				}
			}		
		});
	}

	AddPacket(Name, Function) {
		this.Packets[Name] = Function;
	}
}
module.exports = GSocket;